### PR TITLE
Add api endpoints for documents

### PR DESCRIPTION
**Pull request by 70ray**
This works only with the documents referred to in $external_faq_overrides in pinc/faq.inc but this is sufficient for the present purpose.
Sandbox at: https://www.pgdp.org/~rp31/c.branch/api_docs

### CODE REVIEW

**Review by chrismiceli — APPROVED** [review]

File: api/dp-openapi.yaml
Comment (srjfoo) [inline api/dp-openapi.yaml:1314]: Perhaps it should have been obvious to me, and will be to those who actually use the API to pull the documentation, but I was unsure just what string to use to pull the URL to the documentation when I first started testing. Would it be a good idea to document that the document names should be drawn from the list of documents in the results from the `/documents` query?

Everything else looks good.
---

Specifying a 2-letter language code worked just fine for `/documents`, inasmuch as the list that was returned was limited to only the docs available in that language.

If I had my language pref set for French for `/documents/{document_name}`, `curl` correctly gave me the link to the French version of the proofreading guidelines 🎉 . However, when testing in Swagger, `/documents/{document_name}` does not return the language result--I assume because Swagger doesn't have access to my prefs on TEST.

File: api/v1_docs.inc
Comment (cpeel) [inline api/v1_docs.inc:22]: This function needs to accept a locale so it knows which URL to return. And `get_faq_url()` below _must_ be passed a locale -- either the one specified by the user or `en_US` so that it returns English if none was specified. We should not have generalized endpoints rely on the user's preferences like this one does (and @srjfoo discovered with testing).

File: api/dp-openapi.yaml
Comment (cpeel) [inline api/dp-openapi.yaml:1314]: I would make this `/documents/{document}:` vs `document_name` since we return a filename and not the name of the document itself.

File: api/dp-openapi.yaml
Comment (cpeel) [inline api/dp-openapi.yaml:1303]: This name and description isn't going to be useful, we can help them out:
```suggestion
      - name: language_code
        in: query
        description: Two-letter (ISO 639-1) language code. 
        schema:
          type: string
```

File: api/dp-openapi.yaml
Comment (cpeel) [inline api/dp-openapi.yaml:1297]:
```suggestion
      description: Gets an array of available documents. If language_code is given, only documents available in that language is returned.
```

File: api/dp-openapi.yaml
Comment (cpeel) [inline api/dp-openapi.yaml:1318]:
```suggestion
      description: Gets a URL for the given document in language_code (if specified and available) or English.
```

File: api/dp-openapi.yaml
Comment (cpeel) [inline api/dp-openapi.yaml:1328]:
```suggestion
          description: URL for the document
```

File: api/v1_validators.inc
Comment (cpeel) [inline api/v1_validators.inc:71]: s/document_name/document/ in whole function

File: api/dp-openapi.yaml
Comment (70ray) [reply api/dp-openapi.yaml:1314]: good point.

File: api/v1_docs.inc
Comment (70ray) [reply api/v1_docs.inc:22]: Anyhow, I've changed it as you suggested. It also seems rather unsatisfactory that _GET should be referenced inside the function get_desired_language() but with the language code specified we won't ever get there.

File: api/dp-openapi.yaml
Comment (cpeel) [inline api/dp-openapi.yaml:1306]:
```suggestion
          description: Array of document filenames
```

**Review by cpeel — APPROVED** [review]

**Review by jchaffraix — APPROVED** [review]

File: api/dp-openapi.yaml
Comment (70ray) [reply api/dp-openapi.yaml]: Oops, missed that one.

**Review by srjfoo — APPROVED** [review]

**Comment by srjfoo** [comment]
I need to rescind my approval. There's a disconnect between `lang_code` and `language_code`. 

Last time I tested, both `fr` and `de` gave a subset list of the document filenames  for `/documents`. Now, both queries are giving the full list. In the curl query from the command line, I changed the `language_code` that Swagger uses to `lang_code` and got the correct list. And `/documents/{document}` is still giving me the English  result for a non-English 2-letter language code (also for "language code"_"Country Code").

**Review by srjfoo — CHANGES_REQUESTED** [review]
Please see https://github.com/DistributedProofreaders/dproofreaders/pull/1398#issuecomment-2552747727 -- between the last time I tested, and this time, something broke.

**Comment by 70ray** [comment]
> I need to rescind my approval. There's a disconnect between `lang_code` and `language_code`.
> 
> Last time I tested, both `fr` and `de` gave a subset list of the document filenames for `/documents`. Now, both queries are giving the full list. In the curl query from the command line, I changed the `language_code` that Swagger uses to `lang_code` and got the correct list. And `/documents/{document}` is still giving me the English result for a non-English 2-letter language code (also for "language code"_"Country Code").

Yes, this is inconsistent, I'll change it. There remains the question of what happens if an invalid language code or a valid but unavailable one is given. Should we throw exceptions with messages like "xx is not a valid language code" or "the document is not available in Welsh" or silently return the English one?

**Comment by srjfoo** [comment]
> Yes, this is inconsistent, I'll change it. There remains the question of what happens if an invalid language code or a valid but unavailable one is given. Should we throw exceptions with messages like "xx is not a valid language code" or "the document is not available in Welsh" or silently return the English one?

Right now, English is returned when it doesn't recognize something. I don't really like that as a default, but I'll defer to @cpeel on that.

**Comment by cpeel** [comment]
> > Yes, this is inconsistent, I'll change it. There remains the question of what happens if an invalid language code or a valid but unavailable one is given. Should we throw exceptions with messages like "xx is not a valid language code" or "the document is not available in Welsh" or silently return the English one?
> 
> Right now, English is returned when it doesn't recognize something. I don't really like that as a default, but I'll defer to @cpeel on that.

Yes, throwing an error is probably the right thing to do in the API. Perhaps a change like this (untested):
```diff
diff --git a/pinc/faq.inc b/pinc/faq.inc
index 739d184cd..b3b73b948 100644
--- a/pinc/faq.inc
+++ b/pinc/faq.inc
@@ -63,7 +63,7 @@ $external_faq_overrides = [
 ];


-function get_faq_url(string $faq, ?string $langcode = null): string
+function get_faq_url(string $faq, ?string $langcode = null, bool $english_fallback = True): string
 {
     global $code_dir, $code_url;

@@ -72,8 +72,13 @@ function get_faq_url(string $faq, ?string $langcode = null): string
         $langcode = substr(get_desired_language(), 0, 2);
     }

+    $check_lang_codes = [$langcode];
+    if ($english_fallback) {
+        array_push($check_lang_codes, "en");
+    }
+
     // first try $langcode, then English
-    foreach ([$langcode, "en"] as $langcode) {
+    foreach ($check_lang_codes as $langcode) {
         // Any overrides?
         $external_faq = get_external_faq($faq, $langcode);
         if ($external_faq) {
@@ -87,7 +92,7 @@ function get_faq_url(string $faq, ?string $langcode = null): string
     }

     // fall back to the English version in the code if it exists
-    if (is_file("$code_dir/faq/$faq")) {
+    if ($english_fallback && is_file("$code_dir/faq/$faq")) {
         return "$code_url/faq/$faq";
     }

```

And having the caller in the API check and if `""` is returned raise a `RuntimeException()`.

**Comment by 70ray** [comment]
Changes implemented, Sandbox updated.

**Review by cpeel — APPROVED** [review]

**Review by srjfoo — APPROVED** [review]

**Pull request merged**
